mm2src/mm2_metrics/src/recorder.rs

```diff
@@ -0,0 +1,205 @@
+            }
+            output.push('\n');
+        }
+
+        for (key, histogram) in histograms.drain() {
+            let key = Key::from_name(key.to_owned());
+            let (name, labels) = key_to_parts(&key, None);
+            write_type_line(&mut output, &name, "histogram");
+            for (_, values) in histogram {
+                let mut count = 0;
+                let mut sum = 0.0;
+
+                count += values.len();
+                values.iter().for_each(|value| {
+                    sum += value;
+                });
+
+                write_metric_line::<&str, f64>(&mut output, &name, Some("sum"), &labels, None, sum);
+                write_metric_line::<&str, usize>(
+                    &mut output,
+                    &name,
+                    Some("count"),
+                    key_to_parts(&key, None).1.as_slice(),
+                    None,
+                    count,
+                );
+                output.push('\n');
+            }
+        }
+
+        output
+    }
+
+    pub fn prepare_json(&self) -> MetricsJson {
+        let mut output = vec![];
+
+        for (key, counter) in self.registry.get_counter_handles() {
+            let (key, labels) = key.into_parts();
+            let value = counter.get_inner().load(Ordering::Acquire);
+            output.push(MetricType::Counter {
+                key: key.as_str().to_string(),
+                labels: labels_into_parts(labels.clone().iter()),
+                value,
+            });
+        }
+
+        for (key, gauge) in self.registry.get_gauge_handles() {
+            let (key, labels) = key.into_parts();
+            let value = f64::from_bits(gauge.get_inner().load(Ordering::Acquire));
+            output.push(MetricType::Gauge {
+                key: key.as_str().to_string(),
+                labels: labels_into_parts(labels.clone().iter()),
+                value,
+            });
+        }
+
+        for (key, histogram) in self.registry.get_histogram_handles() {
+            let value = histogram.get_inner().data();
+            let (key, labels) = key.into_parts();
+            let mm_histogram = MmHistogram::new(&value);
+
+            if let Some(qauntiles_value) = mm_histogram {
+                output.push(MetricType::Histogram {
+                    key: key.as_str().to_string(),
+                    labels: labels_into_parts(labels.clone().iter()),
+                    quantiles: qauntiles_value.to_json_quantiles(),
+                });
+            }
+        }
+
+        MetricsJson { metrics: output }
+    }
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+fn key_value_to_snapshot_entry<V>(
+    metrics: &mut HashMap<String, HashMap<Vec<String>, V>>,
+    key: Key,
+    value: V,
+    default: V,
+) {
+    let (key_name, labels) = key_to_parts(&key, None);
+    let entry = metrics
+        .entry(key_name)
+        .or_insert_with(HashMap::new)
+        .entry(labels)
+        .or_insert(default);
+    *entry = value;
```

Comment:
There is no need to take `default` value here as we can rewrite this function by using [insert_entry](https://doc.rust-lang.org/stable/std/collections/hash_map/enum.Entry.html#method.insert_entry).
If the method is not stable yet, please consider matching `Vacant` and `Occupied` variants on the entry.

Reply:
done

Reply:
Sorry, I confused you. There is also [Entry::or_insert_with](https://doc.rust-lang.org/stable/std/collections/hash_map/enum.Entry.html#method.or_insert_with). [Entry::insert_entry](https://doc.rust-lang.org/stable/std/collections/hash_map/enum.Entry.html#method.insert_entry) is even not correct here as it replaces the inner value.
Anyway, it's not critical, but please consider refactoring it if you have time.